gcsio/src/test/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageTest.java

```diff
@@ -3379,6 +3383,94 @@ public void initializeRequest_withDownscopedAccessToken() throws IOException {
         .isEqualTo("Bearer testDownscopedAccessToken");
   }
 
+  @Test
+  public void checkDeletionOrderForHnBucketBalancedFolders() throws InterruptedException {
+    String folderString = "test-folder-start/";
+    List<FolderInfo> foldersToDelete = new LinkedList<>();
+    List<FolderInfo> orderOfDeletion = new LinkedList<>();
+
+    addFolders(foldersToDelete, folderString);
+    int folderSize = foldersToDelete.size();
+    DeleteFolderOperation deleteFolderOperation =
+        new DeleteFolderOperation(foldersToDelete, GoogleCloudStorageOptions.DEFAULT, null);
+
+    deleteFolderOperation.computeChildrenForFolderResource();
+    while (folderSize != 0 && deleteFolderOperation.encounteredNoExceptions()) {
+      FolderInfo folderToDelete = deleteFolderOperation.getElementFromBlockingQueue();
+      orderOfDeletion.add(folderToDelete);
+      folderSize--;
+      deleteFolderOperation.addToToBatchExecutorQueue(() -> null, folderToDelete, 1);
+    }
+    assertThat(orderOfDeletion.size()).isEqualTo(foldersToDelete.size());
+
+    // Map to store the index at which a folder was deleted
+    HashMap<String, Integer> deletionOrder = new HashMap<>();
+    for (int i = 0; i < orderOfDeletion.size(); i++) {
+      deletionOrder.put(orderOfDeletion.get(i).getFolderName(), i);
+    }
+
+    for (int i = 0; i < orderOfDeletion.size(); i++) {
+      FolderInfo curFolder = orderOfDeletion.get(i);
+      String curFolderName = curFolder.getFolderName();
+      String parentFolderName = curFolder.getParentFolderName();
+
+      if (!Strings.isNullOrEmpty(parentFolderName)) {
+        assertThat(deletionOrder.get(parentFolderName) > deletionOrder.get(curFolderName)).isTrue();
```

Comment:
Do you also want to validate that the folder is actually deleted?

Reply:
How would we verify that ? Since this is a UT, we arent creating any actual folder resource and thus neither calling API to delete it. DeletionCallback is returning null followed by the success callback that processes the parent. Is that not sufficient to say that folder was deleted in stimulation ?

Reply:
Why not validate that success callback is received for all the deletes?